### PR TITLE
[Enhancement] Support query dump for materialized view 

### PR DESCRIPTION
**Pull request by LiShuMing**
To support query dump which query's table may contain mv, I did the changes:
- Add "``" for output columns of Materialized View because mv's column may contain complex expressions (eg, `bitmap_union()`;
- Only add dump info into QueryDumpInfo only when it's from HTTP query dump or `enable_query_dump` is true;
- When adding mv into dump info,  optimize the mv again because tables in query may be only a part of mv, and mv may contain other unique/foreign constraints in its properties.
- Make the `tableMap `  in QueryDumpInfo ordered by the insert time because MV's replay needs the basic tables are created.
- Mock some places to support MVs with external tables.

To support dump mv on mv queries, I added two invisible session variables:
- query_excluding_mv_names indicates Query's candidate mvs should exculde.
- query_including_mv_names indicates Query's candidate mvs should include.

I added three cases:
- normal SR tables with MV;
- - normal SR tables with MV on MV;
- external Hive Tables with MV;

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/http/HttpServerHandler.java [hunk 1]
```diff
@@ -116,8 +119,12 @@ private void writeResponse(ChannelHandlerContext context, HttpResponseStatus sta
 
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
-        LOG.warn(cause);
+        StringWriter sw = new StringWriter();
+        PrintWriter pw = new PrintWriter(sw);
+        cause.printStackTrace(pw);
+        LOG.warn(sw.toString());
         ctx.close();
+        throw new Exception(cause);
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/http/HttpServerHandler.java:127]: why throw again?

**Review by liuyehcf — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 79 / 97 (81.44%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiMetadata.java|0|4|00.00%|[173, 174, 175, 176]|
|:large_blue_circle:|com/starrocks/http/rest/QueryDumpAction.java|0|1|00.00%|[92]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|0|1|00.00%|[329]|
|:large_blue_circle:|com/starrocks/http/HttpServerHandler.java|0|2|00.00%|[119, 120]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|8|37.50%|[404, 405, 407, 409, 696]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|3|7|42.86%|[249, 553, 554, 557]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|5|6|83.33%|[680]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetadata.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/dump/QueryDumpInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/InsertAnalyzer.java [hunk 2]
```diff
@@ -243,7 +243,9 @@ public static void analyze(InsertStmt insertStmt, ConnectContext session) {
         insertStmt.setTargetTable(table);
         insertStmt.setTargetPartitionIds(targetPartitionIds);
         insertStmt.setTargetColumns(targetColumns);
-        session.getDumpInfo().addTable(database.getFullName(), table);
+        if (session.getDumpInfo() != null) {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/InsertAnalyzer.java:246]: why check this? 
why not check this before ?

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/InsertAnalyzer.java:246]: Before queries always put the related tables/views/... into DumpInfo.

I've changed this to only save related infos when it's from HTTP query dump or `enable_query_dump` is true.  I think this is more reasonable.

**Review by liuyehcf — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 81 / 104 (77.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiMetadata.java|0|4|00.00%|[173, 174, 175, 176]|
|:large_blue_circle:|com/starrocks/http/rest/QueryDumpAction.java|0|1|00.00%|[92]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|0|1|00.00%|[329]|
|:large_blue_circle:|com/starrocks/http/HttpServerHandler.java|0|5|00.00%|[122, 123, 124, 125, 127]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|8|37.50%|[404, 405, 407, 409, 696]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|3|7|42.86%|[249, 553, 554, 557]|
|:large_blue_circle:|com/starrocks/sql/optimizer/dump/HiveTableDumpInfo.java|2|4|50.00%|[158, 159]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|5|6|83.33%|[680]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetadata.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/dump/QueryDumpInfo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 106 / 126 (84.13%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiMetadata.java|0|4|00.00%|[173, 174, 175, 176]|
|:large_blue_circle:|com/starrocks/http/rest/QueryDumpAction.java|0|1|00.00%|[92]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|0|1|00.00%|[329]|
|:large_blue_circle:|com/starrocks/http/HttpServerHandler.java|0|2|00.00%|[119, 120]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|8|37.50%|[404, 405, 407, 409, 696]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|3|7|42.86%|[249, 553, 554, 557]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|8|75.00%|[2079, 2080]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|5|6|83.33%|[680]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|23|23|100.00%|[]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetadata.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/dump/QueryDumpInfo.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 107 / 127 (84.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hudi/HudiMetadata.java|0|4|00.00%|[173, 174, 175, 176]|
|:large_blue_circle:|com/starrocks/http/rest/QueryDumpAction.java|0|1|00.00%|[92]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|0|1|00.00%|[329]|
|:large_blue_circle:|com/starrocks/http/HttpServerHandler.java|0|2|00.00%|[119, 120]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|8|37.50%|[404, 405, 407, 409, 696]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|3|7|42.86%|[249, 553, 554, 557]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|8|75.00%|[2079, 2080]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|5|6|83.33%|[680]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|23|23|100.00%|[]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetadata.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/OptExternalPartitionPruner.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/CatalogMgr.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|17|17|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/dump/QueryDumpInfo.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|1|1|100.00%|[]|

**Review by murphyatwork — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**